tensorboard/backend/event_processing/data_provider_test.py

```diff
@@ -0,0 +1,162 @@
+        scalar_summary.scalar("square", tf.sign(tf.sin(float(i))), step=i)
+        # Summary with rank-0 data but not owned by the scalars plugin.
+        metadata = summary_pb2.SummaryMetadata()
+        metadata.plugin_data.plugin_name = "marigraphs"
+        tf.summary.write("high_tide", tensor=i, step=i, metadata=metadata)
+
+    logdir = os.path.join(self.logdir, "pictures")
+    with tf.summary.create_file_writer(logdir).as_default():
+      purple = tf.constant([[[255, 0, 255]]], dtype=tf.uint8)
+      for i in xrange(1, 11):
+        image_summary.image("purple", [tf.tile(purple, [i, i, 1])], step=i)
+
+  def create_provider(self):
+    multiplexer = event_multiplexer.EventMultiplexer()
+    multiplexer.AddRunsFromDirectory(self.logdir)
+    multiplexer.Reload()
+    return data_provider.MultiplexerDataProvider(multiplexer)
+
+  def test_list_scalars_all(self):
+    provider = self.create_provider()
+    result = provider.list_scalars(
+        experiment_id="unused",
+        owner_plugin=scalar_metadata.PLUGIN_NAME,
+        run_tag_filter=None,
+    )
+    self.assertItemsEqual(result.keys(), ["polynomials", "waves"])
+    self.assertItemsEqual(result["polynomials"].keys(), ["square", "cube"])
+    self.assertItemsEqual(result["waves"].keys(), ["square", "sine"])
+    sample = result["polynomials"]["square"]
+    self.assertIsInstance(sample, base_provider.ScalarMetadata)
+    self.assertEqual(sample.max_step, 18)
+    # nothing to test for wall time, as it can't be mocked out
+    self.assertIsInstance(sample.summary_metadata, summary_pb2.SummaryMetadata)
+    self.assertEqual(
+        sample.summary_metadata.plugin_data.plugin_name,
+        scalar_metadata.PLUGIN_NAME,
+    )
+
+  def test_list_scalars_filters(self):
+    provider = self.create_provider()
+
+    result = provider.list_scalars(
+        experiment_id="unused",
+        owner_plugin=scalar_metadata.PLUGIN_NAME,
+        run_tag_filter=base_provider.RunTagFilter(tags=["square", "quartic"]),
+    )
+    self.assertItemsEqual(result.keys(), ["polynomials", "waves"])
+    self.assertItemsEqual(result["polynomials"].keys(), ["square"])
+    self.assertItemsEqual(result["waves"].keys(), ["square"])
+
+    result = provider.list_scalars(
+        experiment_id="unused",
+        owner_plugin=scalar_metadata.PLUGIN_NAME,
+        run_tag_filter=base_provider.RunTagFilter(runs=["waves", "hugs"]),
+    )
+    self.assertItemsEqual(result.keys(), ["waves"])
+    self.assertItemsEqual(result["waves"].keys(), ["sine", "square"])
+
+    result = provider.list_scalars(
+        experiment_id="unused",
+        owner_plugin=scalar_metadata.PLUGIN_NAME,
+        run_tag_filter=base_provider.RunTagFilter(["un"], ["likely"]),
+    )
+    self.assertEqual(result, {})
+
+  def test_read_scalars(self):
+    provider = self.create_provider()
+
+    run_tag_filter = base_provider.RunTagFilter(
+        runs=["waves", "polynomials"], tags=["sine", "square", "cube"]
```

Comment:
Maybe add a test for ignoring runs/tags that aren't valid?

Reply:
This does ignore _combinations_ of runs and tags that aren’t valid (no
tag `cube` in run `waves`). Added a completely unused tag and run, too.